here-naksha-lib-core/src/main/java/com/here/naksha/lib/core/AbstractTask.java

```diff
@@ -398,63 +402,61 @@ protected final void unlock() {
-          }
-        }
-        if (!internal && threadCount >= LIMIT) {
+      long myIncrementedValue = AbstractTask.threadCount.updateAndGet(l -> {
+        if (!internal && AbstractTask.threadCount.get() >= LIMIT) {
           log.info(
               "NAKSHA_ERR_REQ_LIMIT_4_INSTANCE - [Request Limit breached for Instance => appId,author,limit,crtValue] - ReqLimitForInstance {} {} {} {}",
               context.getAppId(),
-              author,
+              getAuthorName(),
               LIMIT,
               threadCount);
-          throw new TooManyTasks(LIMIT, 0, true);
+          String errorMessage = "Maximum number of concurrent tasks reached for instance (" + LIMIT + ")";
+          throw new TooManyTasks(errorMessage);
         }
-        if (AbstractTask.threadCount.compareAndSet(threadCount, threadCount + 1)) {
-          if (author != null) {
-            AbstractTask.incrementAuthorUsage(author);
-          }
-          try {
-            state.set(State.START);
-            final Future<RESULT> future = threadPool.submit(this::init_and_execute);
-            return future;
-          } catch (RejectedExecutionException e) {
-            throw new TooManyTasks();
-          } catch (Throwable t) {
-            AbstractTask.threadCount.decrementAndGet();
-            if (author != null) {
-              AbstractTask.decrementAuthorUsage(author);
-            }
-            log.atError()
-                .setMessage("Unexpected exception while trying to fork a new thread")
-                .setCause(t)
-                .log();
-            throw new RuntimeException("Internal error while forking new worker thread", t);
-          }
-        }
-        // Conflict, two threads concurrently try to fork.
-      } while (true);
+        return l + 1;
+      });
+      assert myIncrementedValue >= 0L;
+      String principal = getPrincipalName();
+      if (principal != null) {
+        AtomicInteger authorUsageValue = principalUsageMap.merge(
+            principal, new AtomicInteger(0), (oldAuthorUsageValue, newAuthorUsageValue) -> {
+              int authorLimit = (int) (LIMIT * principalThreshold);
+              if (!internal && oldAuthorUsageValue.get() >= authorLimit) {
+                log.info(
+                    "NAKSHA_ERR_REQ_LIMIT_4_PRINCIPAL - [Request Limit breached for Principal => appId,author,limit,crtValue] - ReqLimitForPrincipal {} {} {} {}",
+                    context.getAppId(),
+                    getAuthorName(),
+                    authorLimit,
+                    oldAuthorUsageValue);
+                AbstractTask.threadCount.decrementAndGet();
+                String errorMessage = "Maximum number of concurrent tasks reached for principal ("
+                    + principalLimit + ")";
+                throw new TooManyTasks(errorMessage);
+              }
+              oldAuthorUsageValue.incrementAndGet();
+              return oldAuthorUsageValue;
+            });
+      }
```

Comment:
The whole function has become bigger with good amount of if-and-else's for someone else to understand clearly what is happening. I suggest modularizing it for better readability.